5.0.0-beta.5 Zod schema rejects RemoteRef-wrapped step inputs from production service

### Environment

    - SDK: workflow@5.0.0-beta.5 (published 2026-05-11)
    - Worker (server-side, per error User-Agent): @workflow/world-vercel/5.0.0-beta.4
    - Runtime: Node 24, Vercel Functions, fluid compute
    - Framework: Next.js 16.2 (Turbopack, React Compiler)
    - Region: fra1
    - Date observed: 2026-05-13, ~04:00 UTC onward

### Summary
Every call to start(myWorkflow, [...]) returns 500 from our API route. The Workflow service accepts the run (a wrun_* ID is allocated and steps fire), but the SDK throws a
    WorkflowWorldError when validating the response to POST /v3/runs/{runId}/events. The service is now returning step inputs wrapped as { _type: 'RemoteRef', _ref: 'dbrf:...' }
    (blob-storage refs), but the SDK's Zod schema requires run.input and step.input as non-optional concrete values.

#### Regression history
    - workflow@4.2.4 (until 2026-05-12 evening UTC): worked.
    - Service-side change rolled out overnight 2026-05-12 -> 13.
    - workflow@4.2.4 after the change: errored on run.output / run.error / run.completedAt being undefined for in-flight runs.
    - Upgraded to workflow@5.0.0-beta.5: now errors on run.input / step.input.

#### Full error (post-upgrade, beta.5)
```
    Error [WorkflowWorldError]: Schema validation failed for POST /v3/runs/wrun_01KRFVDK2PHG87MGD5XQ9WTKJK/events:
      run.input: Invalid input: expected nonoptional, received undefined
      step.input: Invalid input: expected nonoptional, received undefined
```
    Response context:
```
Content-Type: application/cbor, 2816 bytes (CBOR), preview: {
      event: {
        runId: 'wrun_XXXX',
        eventId: 'evnt_XXXX',
        correlationId: 'step_XXXX',
        eventType: 'step_created',
        eventData: {
          stepName: 'step//./workflows/foo-bar.workflow//fooStep',
          input: {
            _type: 'RemoteRef',
            _ref: 'dbrf:team_XXXX:prj_XXXXX:production:wrun_XXX:...'

    [cause] ZodError:
    [
      { code: 'invalid_type', expected: 'nonoptional', path: ['run','input'],  message: 'Invalid input: expected nonoptional, received undefined' },
      { code: 'invalid_type', expected: 'nonoptional', path: ['step','input'], message: 'Invalid input: expected nonoptional, received undefined' }
    ]
```
    Pre-upgrade (4.2.4) variant of the same class:
```
    run.output:      Invalid input: expected nonoptional, received undefined
    run.error:       Invalid input: expected nonoptional, received undefined
    run.completedAt: Invalid input: expected nonoptional, received undefined
```

#### Where it fires

    - Surfaces as Queue callback error on POST /.well-known/workflow/v1/flow for every event the SDK tries to acknowledge.

### Repro

    1. Install workflow@^5.0.0-beta.5 in a Next.js 16 app on Vercel production.
    2. Define a workflow with at least one 'use step' function that takes a non-trivial input (anything that the service decides to externalize as a blob ref — we see it on a few-KB
    structured object).
    3. Call start(workflow, [input]) from an API route.
    4. Observe 500 + the error above; verify the run still exists via the Workflow Run viewer.


### What I think the fix is

    Either:
    - (SDK) Loosen the events-response Zod schema: run.input / step.input / run.output / run.error / run.completedAt should all be .optional() or accept the RemoteRef discriminated union - the service legitimately omits them for in-flight runs and externalizes large payloads.
    - (Service) Stop returning bare RemoteRef for input/output in events that hit pre-5.0 SDKs, or inline payloads under some threshold.

Hitting this from a clean Vercel project today (May 14, 2026). Reproduction is fully blocked on both stable and beta. Adding details in case it helps narrow the regression window.

**Environment**
- Vercel project on a fresh deployment (May 14)
- Build: Nitro 3.0.260429-beta + Vite 5.4.19 (SPA + Express server)
- Runtime: Node 24.x on Vercel Functions, region iad1
- workflow tested at both `4.2.4` (stable) and `5.0.0-beta.5`
- zod tested at `4.4.3` (latest in `^4.3.6` range); workflow's own pin is `4.3.6`

**Trigger**
Standard `start(myWorkflow, [syncLogId])` from an Express route that's mounted via Nitro. The workflow function has 9 step types and uses `Promise.all` to fan out monthly per-service syncs. Input is a single integer (the sync_logs primary key) — not a large payload, so it shouldn't be hitting the RemoteRef externalization path, yet the schema validation still fails on `run.input` and `step.input`.

**v5 beta 5 error**
```
Queue callback error: Error: Schema validation failed for POST /v3/runs/wrun_.../events:
  run.input: Invalid input: expected nonoptional, received undefined
  step.input: Invalid input: expected nonoptional, received undefined

  at createWorkflowRunEvent (file:///var/task/index.mjs:184372:12)
  at trace (file:///var/task/index.mjs:176497:12)

Caused by: ZodError: [
  { "code": "invalid_type", "expected": "nonoptional", "path": ["run","input"], ... },
  { "code": "invalid_type", "expected": "nonoptional", "path": ["step","input"], ... }
]

url: 'https://vercel-workflow.com/api/v3/runs/wrun_.../events'
name: 'WorkflowWorldError'
```

**v4.2.4 stable error (different fields, same shape)**
```
Schema validation failed for POST /v3/runs/wrun_.../events:
  run.output: Invalid input: expected nonoptional, received undefined
  run.error: Invalid input: expected nonoptional, received undefined
  run.completedAt: Invalid input: expected nonoptional, received undefined
```

**Symptoms**
- `start()` returns a `runId` successfully (run gets created on Vercel's side, visible via `npx workflow inspect run <runId>`)
- Run sits in `status: 'pending'` indefinitely
- `/.well-known/workflow/v1/flow` returns 500 on every queue callback attempt
- Run never executes any steps

**Notes**
- Even with my input being trivial (single integer), the `run.input` field is being treated as undefined on the event-creation path. So this isn't *only* about RemoteRef-wrapped large payloads — the schema mismatch shows up for small concrete inputs too.
- Workflow runtime confirms specVersion: 3 in the run record (`npx workflow inspect run <runId>`).

Happy to share more reproduction details if useful.

--- a/.changeset/optional-input-on-snapshots.md
+++ b/.changeset/optional-input-on-snapshots.md
@@ -1,0 +1,6 @@
+---
+"@workflow/world": patch
+"workflow": patch
+---
+
+Make `run.input` and `step.input` `.optional()` on the World snapshot schemas so consumers no longer fail validation when the service externalizes payloads as `RemoteRef` blobs.

--- a/packages/world/src/runs.ts
+++ b/packages/world/src/runs.ts
@@ -47,7 +47,7 @@ export const WorkflowRunBaseSchema = z.object({
   // Optional in database for backwards compatibility, defaults to 1 (legacy) when reading
   specVersion: z.number().optional(),
   executionContext: z.record(z.string(), z.any()).optional(),
-  input: SerializedDataSchema,
+  input: SerializedDataSchema.optional(),
   output: SerializedDataSchema.optional(),
   /**
    * The thrown value from a run_failed event, serialized via the workflow

--- a/packages/world/src/steps.ts
+++ b/packages/world/src/steps.ts
@@ -44,7 +44,7 @@ export const StepSchema = z.object({
    */
   stepName: z.string(),
   status: StepStatusSchema,
-  input: SerializedDataSchema,
+  input: SerializedDataSchema.optional(),
   output: SerializedDataSchema.optional(),
   /**
    * The thrown value from a step_retrying or step_failed event, serialized

